osu.Game/Screens/Play/MasterGameplayClockContainer.cs

```diff
@@ -68,11 +79,33 @@ public MasterGameplayClockContainer(WorkingBeatmap beatmap, double gameplayStart
             firstHitObjectTime = beatmap.Beatmap.HitObjects.First().StartTime;
         }
 
-        [BackgroundDependencyLoader]
-        private void load(OsuConfigManager config)
+        protected override void LoadComplete()
         {
+            base.LoadComplete();
+
             userAudioOffset = config.GetBindable<double>(OsuSetting.AudioOffset);
-            userAudioOffset.BindValueChanged(offset => userOffsetClock.Offset = offset.NewValue, true);
+            userAudioOffset.BindValueChanged(offset => userGlobalOffsetClock.Offset = offset.NewValue, true);
+
+            beatmapOffsetSubscription = realm.RegisterCustomSubscription(r =>
+            {
+                var userSettings = r.Find<BeatmapInfo>(beatmap.BeatmapInfo.ID)?.UserSettings;
+
+                if (userSettings == null) // only the case for tests.
+                    return null;
+
+                void onUserSettingsOnPropertyChanged(object sender, PropertyChangedEventArgs args)
+                {
+                    if (args.PropertyName == nameof(BeatmapUserSettings.Offset))
+                        updateOffset();
+                }
+
+                updateOffset();
+                userSettings.PropertyChanged += onUserSettingsOnPropertyChanged;
+
+                return new InvokeOnDisposal(() => userSettings.PropertyChanged -= onUserSettingsOnPropertyChanged);
+
+                void updateOffset() => userBeatmapOffsetClock.Offset = userSettings.Offset;
+            });
```

Comment:
Definitely can agree that this is a bit gnarly, and the fact that this is duplicated in `BeatmapOffsetControl` verbatim tells me that this sort of boilerplate may start showing up all over soon.

A realm-based bindable may be a decent way forward, but another, perhaps more light-weight alternative would be to just extract most of the `PropertyChanged` event set-up boilerplate. So something like a

```csharp
public IDisposable SubscribeToPropertyChanged<TModel>(Expression<Func<TModel>> modelAccessor, Action<TModel> onChanged);
```

wherein you'd fetch the name of the property to check for in the `PropertyChangedEventArgs` from the expression (probably without nested support for now, just access the outermost `MemberExpression` and pull a name from it). But the above is pretty constraining...

Reply:
Can you pull out the method name successfully by using an `Expression`? That may solve the issues which caused my WIP bindable implementation to be a bit ugly:

```csharp
// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
// See the LICENCE file in the repository root for full licence text.

using System;
using System.ComponentModel;
using System.Diagnostics;
using osu.Framework.Allocation;
using osu.Framework.Bindables;
using osu.Game.Database;
using Realms;

#nullable enable

namespace osu.Game.Screens.Play
{
    /// <summary>
    /// A bindable which tracks a realm based property.
    /// </summary>
    /// <typeparam name="T">The type of our stored <see cref="Bindable{T}.Value"/>.</typeparam>
    public class RealmBindable<T> : Bindable<T>
    {
        private readonly RealmAccess realm;
        private readonly Func<RealmObject, T> propertyLookup;
        private readonly string? propertyName;

        private RealmObject? target;

        private readonly IDisposable subscription;

        public RealmBindable(RealmAccess realm, Func<Realm, RealmObject> instanceLookup, Func<RealmObject, T> propertyLookup, string? propertyName)
        {
            this.realm = realm;
            this.propertyLookup = propertyLookup;
            this.propertyName = propertyName;

            subscription = realm.RegisterCustomSubscription(r =>
            {
                target = instanceLookup(r);
                target.PropertyChanged += onRealmChanged;

                return new InvokeOnDisposal(() => target.PropertyChanged -= onRealmChanged);
            });
        }

        private void onRealmChanged(object sender, PropertyChangedEventArgs args)
        {
            Debug.Assert(target != null);

            if (string.IsNullOrEmpty(propertyName) || args.PropertyName == propertyName)
            {
                Value = propertyLookup(target);
            }
        }

        public override void UnbindFrom(IUnbindable them)
        {
            base.UnbindFrom(them);
            subscription?.Dispose();
        }
    }
}
```

Reply:
It is a bit ad-hoc but possible, yeah. [Here's a sharplab with the POC.](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQcAMACOSB0AMgJZQCO+AogB4AOYApgM5PED2UTA3OunAMy4kANlwJsAYXQBvdNnliEchbLQL1QgJwAKAOINgAWQYBbAEYMwAOQCGJhgB4AIjeAMAKsXsAabKWAAfNoAJtgAvAHYwfgAEmwQYACUiTxqGvJ4OvpGphbWdo4ubp4+flCBIeGR0UUMsfFJKcryAL7N2O30xABurgxCong42cbmlrb2Du6GbMEMADa+7gAKYGy0lsAAnkE09MysHA5wAKxTM3OL2CtrG2DbAZEMdIws7FCJ7arp8sQAZthtABCbTPfZvDh+JjYQh2MzBGx7V6HKDYeZwhHJdo/bDAAAWawA7tgoAxiQBBMAAcwg9nKNAAxgxaMB3toACQAIgA8hA3GATGwmMBsGDke9sCYIMLsBZsDY0RiFWKDu9Ob4oAU2H9QS9VRxkqkcdiFP9ASD0eYEfgAEKzLZQ7CjPJI/Wo+xjMCuiEfT5pHHyfFEklk7CUml04CM5msjgczmwq0Ksz2yXSkVyhUevLyhlMliivU+9UkrU6lU+w0m+TV7AAejr125Tm5IBDwoYoRseYO2AZNlRcqYEFotDY907vkJxHxkvH/RnllcErxlgYtYbgNJndxbD8/OX3QWDqY2uAAFp+zKAAbZyzelFUIso69lYB76/O+/P97X6HQVl5mwa9LXhGx8GWGwwAKfkmAAbUwABdV9iGhPEZz9HEzW0O8vR/Dgn3BFFsCBMJFSTCCoJgyx4KQzCA0DAk2GJUkKWpWkGHpagmRZNkuSsZg3C7HsCwrFFoSg/ooDYEVh1HcchJLTV7G1XUiPeKt/Q0Ws4AAdklXJv3UgivzAfAJgYI0FDaNAWiAA)

Reply:
Is [this](https://github.com/ppy/osu/compare/master...peppy:realm-property-watching?expand=1) what you were thinking? Not sure how legible the invocation is...